libyul/ASTNodeRegistry.h

```diff
@@ -0,0 +1,84 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <map>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace solidity::yul
+{
+
+class ASTNodeRegistry
```

Comment:
You described these classes in the PR description, but I'd much rather just have docstrings for them :)

Reply:
Especially `m_idToLabelMapping` and `m_labels` should be documented. Like, what are the numbers stored in `m_idToLabelMapping` and the assumptions about them (can there be duplicates? do both vectors have to have same length?)

Reply:
Ghost nodes are added during CFG construction. They only live in the CFG itself and are not actually referenced in the AST. We could also remove them here and specialize them out for CFGs. Then it's more local to where they are introduced and needed.

Reply:
Well, yeah, their presence here threw me off because I initially thought you needed them for something in the AST and the naming from CFG was only a coincidence. So might be clearer to separate those.

On the other hand, now that it's properly documented and it's clear why we need it, I'd also be fine with leaving it here. It leaks CFG implementation details a bit, but I imagine that separating this out might complicate things.

Reply:
Yeah it would complicate things to completely separate it out. But it does work well to let the CFG render ghost labels so at least that detail doesn't leak.